dist/angular-promise-buttons.js

```diff
@@ -23,47 +23,110 @@ angular.module('angularPromiseButtons')
+                    }, waitTime);
+                }
+
+                function setFinishedState(btnEl, onEndConfig)
+                {
+                    if (onEndConfig)
+                    {
+                        if (onEndConfig.resultHtml)
+                        {
+                            btnEl.html(onEndConfig.resultHtml);
+                        }
+                        if (onEndConfig.resultCssClass)
+                        {
+                            btnEl.addClass(onEndConfig.resultCssClass);
+                        }
+                    }
                 }
 
-                function handleLoadingFinished(btnEl)
+                function removeLoadingState(btnEl)
                 {
                     if (cfg.btnLoadingClass) {
                         btnEl.removeClass(cfg.btnLoadingClass);
                     }
+                }
+
+                function revertToNormalState(btnEl, defaultHtml, onEndConfig)
+                {
                     if (cfg.disableBtn) {
                         btnEl.removeAttr('disabled');
                     }
+                    if (defaultHtml) {
+                        btnEl.html(defaultHtml);
+                    }
+                    if (onEndConfig && onEndConfig.resultCssClass)
+                    {
+                        btnEl.removeClass(onEndConfig.resultCssClass);
+                    }
                 }
 
                 function initPromiseWatcher(watchExpressionForPromise, btnEl)
                 {
                     // watch promise to resolve or fail
                     scope.$watch(watchExpressionForPromise, function (mVal)
                     {
+                        var initPromise = null;
                         // for regular promises
-                        if (mVal && mVal.then) {
-                            handleLoading(btnEl);
-                            mVal.finally(function ()
-                            {
-                                handleLoadingFinished(btnEl);
-                            });
+                        if (mVal && mVal.then)
+                        {
+                            initPromise = mVal;
                         }
                         // for $resource
-                        else if (mVal && mVal.$promise) {
-                            handleLoading(btnEl);
-                            mVal.$promise.finally(function ()
-                            {
-                                handleLoadingFinished(btnEl);
-                            });
+                        else if (mVal && mVal.$promise)
+                        {
+                            initPromise = mVal.$promise;
+                        }
+
+                        if (initPromise)
+                        {
+                            var defaultHtml = cfg.defaultHtml || btnEl.html();
+                            setLoadingState(btnEl);
+                            initPromise.then(
```

Comment:
They should be. 

If you look at the interface for a promise from angular.d.ts, it shows that .then, .catch, and finally are always defined. 

interface IPromise<T> {
        /**
         * Regardless of when the promise was or will be resolved or rejected, then calls one of the success or error callbacks asynchronously as soon as the result is available. The callbacks are called with a single argument: the result or rejection reason. Additionally, the notify callback may be called zero or more times to provide a progress indication, before the promise is resolved or rejected.
         * The successCallBack may return IPromise<void> for when a $q.reject() needs to be returned
         * This method returns a new promise which is resolved or rejected via the return value of the successCallback, errorCallback. It also notifies via the return value of the notifyCallback method. The promise can not be resolved or rejected from the notifyCallback method.
         */
        then<TResult>(successCallback: (promiseValue: T) => IHttpPromise<TResult>|IPromise<TResult>|TResult|IPromise<void>, errorCallback?: (reason: any) => any, notifyCallback?: (state: any) => any): IPromise<TResult>;

        /**
         * Shorthand for promise.then(null, errorCallback)
         */
        catch<TResult>(onRejected: (reason: any) => IHttpPromise<TResult>|IPromise<TResult>|TResult): IPromise<TResult>;

        /**
         * Allows you to observe either the fulfillment or rejection of a promise, but to do so without modifying the final value. This is useful to release resources or do some clean-up that needs to be done whether the promise was rejected or resolved. See the full specification for more information.
         *
         * Because finally is a reserved word in JavaScript and reserved keywords are not supported as property names by ES3, you'll need to invoke the method like promise['finally'](callback) to make your code IE8 and Android 2.x compatible.
         */
        finally(finallyCallback: () => any): IPromise<T>;
    }